lib/utils/promote.ts

```diff
@@ -117,26 +109,29 @@ async function configure(deviceIp: string, config: any): Promise<void> {
 	const json = JSON.stringify(config);
 	const b64 = Buffer.from(json).toString('base64');
 	const str = `"$(base64 -d <<< ${b64})"`;
-	await execCommand(deviceIp, `os-config join '${str}'`, 'Configuring...');
+	await execCommand(deviceIp, `os-config join ${str}`, 'Configuring...');
 }
 
 async function deconfigure(deviceIp: string): Promise<void> {
 	await execCommand(deviceIp, 'os-config leave', 'Configuring...');
 }
 
 async function assertDeviceIsCompatible(deviceIp: string): Promise<void> {
-	const { exitWithExpectedError } = await import('../utils/patterns');
 	try {
 		await execBuffered(deviceIp, 'os-config --version');
 	} catch (err) {
-		exitWithExpectedError(stripIndent`
+		if (err instanceof ExpectedError) {
+			throw err;
+		}
+		console.error(`${err}\n`);
+		throw new ExpectedError(stripIndent`
 			Device "${deviceIp}" is incompatible and cannot join or leave an application.
-			Please select or provision device with balenaOS newer than ${MIN_BALENAOS_VERSION}.`);
+			Please select or provision a device with balenaOS newer than ${MIN_BALENAOS_VERSION}.`);
 	}
 }
 
 async function getDeviceType(deviceIp: string): Promise<string> {
-	const output = await execBuffered(deviceIp, 'cat /etc/os-release');
+	const output = await getDeviceOsRelease(deviceIp);
```

Comment:
During debugging, I observed that `balena join` was invoking `ssh` twice to run `cat /etc/os-release`. So I created `getDeviceOsRelease` using lodash's memoize to cache the result.